Tasks/Npm/npm.ts

```diff
@@ -0,0 +1,273 @@
+
+        let url = `${collectionUrl}_packaging/${feedId}/npm/registry/`;
+        let trimmedUrl = NpmRegistry._trimRegistryUrl(url);
+        let auth = `${trimmedUrl}:_authToken=${NpmRegistry._getSystemAccessToken()}`;
+
+        return new NpmRegistry(url, auth, authOnly);
+    }
+
+    private static _trimRegistryUrl(registryUrl: string): string {
+        if (registryUrl) {
+            let slashIndex = registryUrl.indexOf(":");
+            return registryUrl.substring(slashIndex + 1);
+        }
+    }
+
+    private static _getSystemAccessToken(): string {
+        let auth = tl.getEndpointAuthorization("SYSTEMVSSCONNECTION", false);
+        if (auth.scheme === "OAuth") {
+            tl.debug(tl.loc("FoundBuildCredentials"));
+            return auth.parameters["AccessToken"];
+        } else {
+            tl.warning(tl.loc("NoBuildCredentials"));
+        }
+
+        return undefined;
+    }
+}
+
+async function main(): Promise<void> {
+    tl.setResourcePath(path.join(__dirname, 'task.json'));
+
+    let command = tl.getInput(NpmTaskInput.Command);
+    let workingDir = tl.getInput(NpmTaskInput.WorkingDir);
+
+    let findNpmrc = workingDir ? tl.findMatch(workingDir, ".npmrc") : [];
+    if (findNpmrc && findNpmrc.length > 1) {
+        throw Error(tl.loc("MultipleProjectConfigs"));
+    }
+    let projectNpmrc = findNpmrc[0];
+    pushNpmrc(projectNpmrc);
+
+    let npmrc = getTempNpmrcPath();
+    let npm = new NpmToolRunner(workingDir, npmrc);
+
+    let endpointId: string;
+    let feedId: string;
+    let registryLocation: string;
+    let npmRegistries: INpmRegistry[] = getLocalNpmRegistries(workingDir);
+    switch(command) {
+        case "install":
+            registryLocation = tl.getInput(NpmTaskInput.CustomRegistry);
+            endpointId = tl.getInput(NpmTaskInput.CustomEndpoint);
+            feedId = tl.getInput(NpmTaskInput.CustomFeed);
+            npm.line("install");
+            break;
+
+        case "publish":
+            registryLocation = tl.getInput(NpmTaskInput.PublishRegistry);
+            endpointId = tl.getInput(NpmTaskInput.PublishEndpoint);
+            feedId = tl.getInput(NpmTaskInput.PublishFeed);
+
+            // We delete their project .npmrc so it won't override our user .npmrc
+            tl.rmRF(`${workingDir}/.npmrc`);
+            npm.line("publish");
+            break;
+
+        case "custom":
+            registryLocation = tl.getInput(NpmTaskInput.CustomRegistry);
+            endpointId = tl.getInput(NpmTaskInput.CustomEndpoint);
+            feedId = tl.getInput(NpmTaskInput.CustomFeed);
+            npm.line(tl.getInput(NpmTaskInput.CustomCommand, true));
+            break;
+
+        default:
+            tl.setResult(tl.TaskResult.Failed, tl.loc("UnknownCommand", command));
+            return;
+    }
+
+    if (registryLocation == RegistryLocation.Feed) {
+        npmRegistries.push(NpmRegistry.FromFeedId(feedId));
```

Comment:
If I understand correctly, we only want to append the registries if they already are in the npmrc.